### DISCUSSION TITLE
Struggle with heat pump values

### DISCUSSION THREAD

**Discussion by gregorlein** [oemof/tespy#488] [Q&A] [answered]
Hi,
I'm trying to set up a model as in the following picture:
![system drawio](https://github.com/oemof/tespy/assets/159910242/bf03fe1c-a08b-4b3d-8b79-fafd909acaf9)

I'm setting it up component by component, but have troubles with the heat pump. I added sources and sinks in the connection 11, 12, 17 and 18 to look only at the heat pump. Where in my code do I have linear dependencies? Or are the provided parameters too inaccurate? I checked the enthalpy-pressure diagram for NH3 and thought the values make sense...

One additional question regarding my system above: Do I need cycle closers in the cycles 9-10-11-12 and 15-16-17-18?

Thanks for your help!

Code:

```python
cd = Condenser('Condenser')
va = Valve('Expnsion Valve')
cp = Compressor('Compressor')
ev = HeatExchanger('Evaporator')
cc_hp = CycleCloser('CC Heat Pump')

so_cd = Source('so_cd')
si_cd = Sink('si_cd')
so_ev = Source('so_ev')
si_ev = Sink('si_ev')

c20 = Connection(ev, 'out2', cc_hp, 'in1', label='20')
c20a = Connection(cc_hp, 'out1', cp, 'in1', label='20a')
c21 = Connection(cp, 'out1', cd, 'in1', label='21')
c22 = Connection(cd, 'out1', va, 'in1', label='22')
c23 = Connection(va, 'out1', ev, 'in2', label='23')

c11 = Connection(so_cd, 'out1', cd, 'in2', label='11a')
c12 = Connection(cd, 'out2', si_cd, 'in1', label='12')
c17 = Connection(so_ev, 'out1', ev, 'in1', label='17')
c18 = Connection(ev, 'out1', si_ev, 'in1', label='18')

nwtest = Network(T_unit='C', p_unit='bar', h_unit='kJ / kg', m_unit = 'kg / s')
nwtest.add_conns(c20, c20a, c21, c22, c23, c11, c12, c17, c18)

cd.set_attr(pr1=0.98, pr2=0.98)
ev.set_attr(pr1=0.98, pr2=0.98)
cp.set_attr(eta_s=0.85)

c20.set_attr(x=1, fluid={'NH3': 1}, T=35) # h=1700 from diagram
c22.set_attr(x=0, T=63) #h=600

c11.set_attr(fluid={'water': 1}, T=65, v=700/3600)
c12.set_attr(T=72)
c17.set_attr(fluid={'water': 1}, T=37, v=700/3600)
c18.set_attr(T=30)

nwtest.solve('design')
nwtest.print_results()
```

**Comment by fwitte** [comment, marked as answer]
@gregorlein,

thanks for reaching out. There is three things in your setup:

1. The `Condenser` class has inbuilt `x=0` at hot side outlet (`out1`), therefore you cannot specify `x=0` there.
2. You fix flow and temperature change on both sides of the system (source and sink side). That cannot work, remove one specification.
3. Instead of those specifications pressure information is missing on source and sink side, add one value respectively, e.g.:

```python
from tespy.components import Condenser, Valve, Compressor, HeatExchanger, CycleCloser, Source, Sink
from tespy.connections import Connection
from tespy.networks import Network


cd = Condenser('Condenser')
va = Valve('Expnsion Valve')
cp = Compressor('Compressor')
ev = HeatExchanger('Evaporator')
cc_hp = CycleCloser('CC Heat Pump')

so_cd = Source('so_cd')
si_cd = Sink('si_cd')
so_ev = Source('so_ev')
si_ev = Sink('si_ev')

c20 = Connection(ev, 'out2', cc_hp, 'in1', label='20')
c20a = Connection(cc_hp, 'out1', cp, 'in1', label='20a')
c21 = Connection(cp, 'out1', cd, 'in1', label='21')
c22 = Connection(cd, 'out1', va, 'in1', label='22')
c23 = Connection(va, 'out1', ev, 'in2', label='23')

c11 = Connection(so_cd, 'out1', cd, 'in2', label='11a')
c12 = Connection(cd, 'out2', si_cd, 'in1', label='12')
c17 = Connection(so_ev, 'out1', ev, 'in1', label='17')
c18 = Connection(ev, 'out1', si_ev, 'in1', label='18')

nwtest = Network(T_unit='C', p_unit='bar', h_unit='kJ / kg', m_unit = 'kg / s')
nwtest.add_conns(c20, c20a, c21, c22, c23, c11, c12, c17, c18)

cd.set_attr(pr1=0.98, pr2=0.98)
ev.set_attr(pr1=0.98, pr2=0.98)
cp.set_attr(eta_s=0.85)

c20.set_attr(x=1, fluid={'NH3': 1}, T=35) # h=1700 from diagram  this one lower than 17/18!!
c22.set_attr(T=63) #h=600  # this temperature must be higher than 11/12!!

c11.set_attr(fluid={'water': 1}, p=1, T=65, v=700/3600)
c12.set_attr(T=72)
c17.set_attr(fluid={'water': 1}, p=1, T=37) # ,v=700/3600)
c18.set_attr(T=30)

nwtest.solve('design')
nwtest.print_results()
```
Btw. the temperature levels for condensation and evaporation do not check out, the condensation temperature must be higher than the respective cold part and in the evaporator its the other way round (see warnings after simulation).

To your second question: I do not think you will need a `CycleCloser` in the extra branches. Try it out with one or the other side. What do you want to do there? Mix down/up temperature level?

**Reply by gregorlein** [reply]
thanks very much, makes all sense now!

Regarding the extra branches: I tried it without CycleCloser, and it does work. In the end I want to use the model with data for the heat demand on the district heating side. In case of low demands e.g. in summer, the source (a geothermal well) provides enough heat and the heat pump is not needed. So I wanted to build in a bypass... Does that sound reasonable?

**Reply by gregorlein** [reply]
Hi @fwitte,
I added a second consumer to my system:
![system drawio](https://github.com/oemof/tespy/assets/159910242/d08eee71-3e57-44b7-b006-9e8ad0f28f7a)

I want to use the demands of both consumers as well as the flow and feedflow and backflow temperatures of the geothermal source as input. The Heat Pump should produce any additional heat needed and the model should adjust the flows in the system according to the demands. With the inputs I use, I run into dependency problems. Does the system not make sense the way I set it up?

Regarding the dashed lines: Later, a storage system should be added where heat is stored in case the demands are below what the geothermal source can provide...

Thanks for any advice!

```python
Code:
from tespy.components import *
from tespy.connections import Connection
from tespy.networks import Network
import CoolProp.CoolProp as CP

dap_source = Source('DAP extraction')
dap_sink = Sink('DAP injection')
dap_pump = Pump('DAP loop Pump')
dap_feed = Pipe('Feed Pipe DAP')
dap_return = Pipe('Return Pipe DAP')
dap_he = HeatExchanger('DAP heat exchanger')

tud = SimpleHeatExchanger('TUD')
tud_pump = Pump('TUD loop pump')
tud_he = HeatExchanger('TUD heat exchanger')
tud_cc = CycleCloser('CC TUD')

cd = Condenser('Condenser')
va = Valve('Expansion Valve')
cp = Compressor('Compressor')
ev = HeatExchanger('Evaporator')
cc_hp = CycleCloser('CC Heat Pump')

owd = SimpleHeatExchanger('OWD')
va_owd = Valve('Valve OWD')
sp_owd = Splitter('Splitter OWD')
me_owd = Merge('Merge OWD')

cc = CycleCloser('CC outer circle')
pu1 = Pump('Pump outer circle')

sp_cd = Splitter('Splitter Condensor')
me_cd = Merge('Merge Condensor')
va_cd = Valve('Valve Condensor')

sp_ev = Splitter('Splitter Evaporator')
me_ev = Merge('Merge Evaporator')
va_ev = Valve('Valve Evaporator')

c0 = Connection(dap_source, 'out1', dap_pump, 'in1', label='0')
c1 = Connection(dap_pump, 'out1', dap_feed, 'in1', label = '1')
c2 = Connection(dap_feed, 'out1', dap_he, 'in1', label='2')
c3 = Connection(dap_he, 'out1', dap_return, 'in1', label='3')
c4 = Connection(dap_return, 'out1', dap_sink, 'in1', label='4')

c5 = Connection(tud, 'out1', tud_cc, 'in1', label='5')
c5a = Connection(tud_cc, 'out1', tud_pump, 'in1', label='5a')
c6 = Connection(tud_pump, 'out1', tud_he, 'in2', label='6')
c7 = Connection(tud_he, 'out2', tud, 'in1', label='7')

c24 = Connection(sp_owd, 'out2', va_owd, 'in1', label='24')
c25 = Connection(va_owd, 'out1', owd, 'in1', label='25')
c26 = Connection(owd, 'out1', me_owd, 'in2', label='26')

c27 = Connection(ev, 'out2', cc_hp, 'in1', label='27')
c27a = Connection(cc_hp, 'out1', cp, 'in1', label='27a')
c28 = Connection(cp, 'out1', cd, 'in1', label='28')
c29 = Connection(cd, 'out1', va, 'in1', label='29')
c30 = Connection(va, 'out1', ev, 'in2', label='30')

c9 = Connection(sp_cd, 'out1', va_cd, 'in1', label='9')
c10 = Connection(va_cd, 'out1', me_cd, 'in1', label='10')
c11 = Connection(sp_cd, 'out2', cd, 'in2', label='11')
c12 = Connection(cd, 'out2', me_cd, 'in2', label='12')

c17 = Connection(sp_ev, 'out1', va_ev, 'in1', label='17')
c18 = Connection(va_ev, 'out1', me_ev, 'in1', label='18')
c19 = Connection(sp_ev, 'out2', ev, 'in1', label='19')
c20 = Connection(ev, 'out1', me_ev, 'in2', label='20')

c8 = Connection(dap_he, 'out2', sp_cd, 'in1', label='8')
c13 = Connection(me_cd, 'out1', sp_owd, 'in1', label='13')
c14 = Connection(sp_owd, 'out1', tud_he, 'in1', label='14')
c15 = Connection(tud_he, 'out1', me_owd, 'in1', label='15')
c16 = Connection(me_owd, 'out1', sp_ev, 'in1', label='16')
c21 = Connection(me_ev, 'out1', pu1, 'in1', label='21')
c22 = Connection(pu1, 'out1', cc, 'in1', label ='22')
c22a = Connection(cc, 'out1', dap_he, 'in2', label='22a')

nw = Network(T_unit='C', p_unit='bar', h_unit='kJ / kg', m_unit = 'kg / s')
nw.add_conns(c0, c1, c2, c3, c4, c5, c5a, c6, c7, c8, c9, c10, c11, c12, c13, c14,
             c15, c16, c17, c18, c19, c20, c21, c22, c22a, c24, c25, c26, c27, c27a, c28, c29, c30)

dap_he.set_attr(pr1=0.98, pr2=0.98, ttd_u=5, ttd_l=5,
            design=['pr1', 'pr2', 'ttd_u'], offdesign=['zeta1', 'zeta2', 'kA_char'])
tud_he.set_attr(pr2=0.98, pr1=0.98, ttd_l=5, ttd_u=5,
            design=['pr1', 'pr2', 'ttd_u'], offdesign=['zeta1', 'zeta2', 'kA_char'])
tud.set_attr(pr=0.95, Q=-10e6, design=['pr'], offdesign=['zeta', 'kA_char'])
owd.set_attr(pr=0.98, Q=-10e6, design=['pr'], offdesign=['zeta', 'kA_char'])

cd.set_attr(pr1=0.98, pr2=0.98, ttd_l=5) 
ev.set_attr(pr1=0.98, pr2=0.98)
cp.set_attr(eta_s=0.85)

dap_pump.set_attr(eta_s=0.75, design=['eta_s'], offdesign=['eta_s_char'])
tud_pump.set_attr(eta_s=0.75, design=['eta_s'], offdesign=['eta_s_char'])
pu1.set_attr(eta_s=0.75,design=['eta_s'], offdesign=['eta_s_char'])

c0.set_attr(fluid={'water': 1}, v = 350/3600, T = 72, p=1, offdesign=['T'])
c3.set_attr(T = 30, offdesign = ['T'])
c4.set_attr(p=1)

c5.set_attr(fluid={'water': 1}, p=2)
c7.set_attr(T=80)

t_hs_bf = 20
c8.set_attr(fluid={'water': 1}, p=2)
c20.set_attr(T=t_hs_bf)

p_eva = CP.PropsSI("P", "Q", 1, "T", t_hs_bf - 5 + 273.15, 'NH3') * 1e-5
c27.set_attr(p=p_eva, x=1, fluid={'NH3':1})

t_owd_ff = 85
p_cond = CP.PropsSI("P", "Q", 0, "T", t_owd_ff + 5 + 273.15, 'NH3') * 1e-5
c29.set_attr(p=p_cond)

dap_feed.set_attr(Q = -250, pr = 0.98, ks = 0.0005, L = 3000, D = "var")
dap_return.set_attr(Q = -250, pr=0.98, ks = 0.0005, L = 3000, D = "var")

nw.solve('design')
nw.print_results()
```

**Reply by fwitte** [reply]
@gregorlein,

I'd invite you to join the online user meeting on Monday (https://tespy.readthedocs.io/en/main/regular_meeting.html) so we can discuss these things. There are a couple of things you will have to take into account in general:

- If parts of the network are shut down (no mass flow), the solver does not work. When such things can happen, I usually use multiple networks, which are simulated depending on the boundary conditions.
- You have created a quite large network in a single step. I usually recommend building (and making it work stepwise) and putting that into some kind of class of function. Then I call that function to generate the network with stable starting values and after that I modify boundary conditions, for example heat demand, inflow or feed flow temperatures etc.. 

I can have a closer look at the scripts when I have some spare time...

Best, have a nice weekend!

Francesco

**Reply by gregorlein** [reply]
Hi Francesco,
yes, I wanted to join the meeting, unfortunately I can't make it today... I will try next month! But thanks for the suggestions, it probably makes sense to have several networks for different modes like high demand, low demand etc.
Kind regards,
Paula

**Comment by fwitte** [comment]
OT: seems like you draw these figures using draw.io (or similar)? Do you have a template for the components and would you like to share it? In case it is draw.io one could potentially also leverage the .json structure of the diagram to generate a tespy network... What do you think?

**Reply by gregorlein** [reply]
yes, I made a library in draw.io with the basic components. I'm new to github... what would be the best way to share it?

**Reply by fwitte** [reply]
That's a good question, maybe it would be cool to start a new project, where the draw.io components are stored and users can use those to also generate the tespy network, components and connections based on its structure... What you could do is just start your own repository and upload the data and describe steps how to use the library in draw.io.
